reactive/openstack.py

```diff
@@ -40,14 +41,24 @@ def upgrade_charm():
     clear_flag('charm.openstack.creds.set')
 
 
+@hook('update-status')
+def update_status():
+    # need to recheck creds in case the credentials from Juju have changed
+    clear_flag('charm.openstack.creds.set')
+
+
 @hook('pre-series-upgrade')
 def pre_series_upgrade():
     layer.status.blocked('Series upgrade in progress')
 
 
 @when_not('charm.openstack.creds.set')
 def get_creds():
+    prev_creds = layer.openstack.get_user_credentials()
     toggle_flag('charm.openstack.creds.set', layer.openstack.get_credentials())
+    creds = layer.openstack.get_user_credentials()
+    if creds != prev_creds:
+        set_flag('charm.openstack.creds.changed')
```

Comment:
I'm not sure how this is supposed to work?  The `get_user_credentials()` function doesn't cache anything in stored state, so it will always get whatever is in the config and / or `credential-get`, so AFAICT these two calls will always be equal and the flag will never be set.

Reply:
> The get_user_credentials() function doesn't cache anything in stored state

Yes it does, see:
* [get_user_credentials](https://github.com/charmed-kubernetes/charm-openstack-integrator/blob/244b8d533f38f18fc6908dc8f9327c9b767a47cc/lib/charms/layer/openstack.py#L125-L126)
* [_load_creds](https://github.com/charmed-kubernetes/charm-openstack-integrator/blob/244b8d533f38f18fc6908dc8f9327c9b767a47cc/lib/charms/layer/openstack.py#L276-L277)

And where the stored value is updated:
* [get_credentials](https://github.com/charmed-kubernetes/charm-openstack-integrator/blob/244b8d533f38f18fc6908dc8f9327c9b767a47cc/lib/charms/layer/openstack.py#L109)

Yep, it's confusing: get_credentials updates the stored credentials, get_user_credentials returns what is stored. This handler has a call to `get_credentials()` sandwiched inbetween the two calls to `get_user_credentials()`, hence the two calls could return different values.

Reply:
Oof, I completely missed that.  Shame on whoever named those functions! :wink: